ocaml-lsp-server/src/ocamlformat_rpc.ml

```diff
@@ -0,0 +1,172 @@
+      (pid, stdout_i, stdin_o)
+    in
+    let open Fiber.O in
+    let input = Unix.in_channel_of_descr stdout in
+    let output = Unix.out_channel_of_descr stdin in
+    let* io_thread =
+      (* Don't create a new thread if there is already one. *)
+      (* This can happen if the Ocamlformat server crashed. *)
+      match io_thread with
+      | Some io_thread -> Fiber.return io_thread
+      | None -> Scheduler.create_thread ()
+    in
+    let* client = pick_client ~pid input output io_thread in
+    match client with
+    | Error (`Msg msg) ->
+      (* The process did start but something went wrong when negociating the
+         version so we need to kill it *)
+      Unix.kill pid Sys.sigkill;
+      Scheduler.stop io_thread;
+      Fiber.return @@ Error (`Msg msg)
+    | Ok client ->
+      let process = { pid; input; output; io_thread; client } in
+      let+ () = configure process in
+      Log.log ~section:"ocamlformat" (fun () ->
+          Log.msg "Ocamlformat-RPC server started" [ ("pid", `Int pid) ]);
+      t := Running process;
+      Ok process)
+
+let start ?io_thread t =
+  let open Fiber.O in
+  try
+    let+ result = start_exn ?io_thread t in
+    match result with
+    | Error (`Msg msg) ->
+      Log.log ~section:"ocamlformat" (fun () ->
+          Log.msg "An error occured while initializing ocamlformat"
+            [ ("msg", `String msg) ]);
+      Error `No_process
+    | r -> r
+  with
+  | Unix.Unix_error (e, _, _) ->
+    t := Errored;
+    Log.log ~section:"ocamlformat" (fun () ->
+        Log.msg "An error occured when starting ocamlformat"
+          [ ("msg", `String (Unix.error_message e)) ]);
+    Fiber.return @@ Error `No_process
+
+let get_process t =
+  let open Fiber.O in
+  match !t with
+  | Stopped -> start t
+  | Running ({ pid; io_thread; _ } as p) -> (
+    (* Is the process still running ? *)
+    match
+      Scheduler.async io_thread (fun () -> Unix.waitpid [ WNOHANG ] pid |> fst)
+    with
+    | Error `Stopped -> Fiber.return @@ Error (`Msg "Thread stopped")
+    | Ok res -> (
+      let* res = Scheduler.await_no_cancel res in
+      match res with
+      | Error e -> Exn_with_backtrace.reraise e
+      | Ok 0 -> Fiber.return @@ Ok p
+      | Ok _ ->
+        Log.log ~section:"ocamlformat" (fun () ->
+            Log.msg "Ocamlformat-RPC server exited" [ ("pid", `Int pid) ]);
+        start ~io_thread t))
+  | Errored
+  | Binary_not_found ->
+    Fiber.return @@ Error `No_process
+
+let query_type t typ =
+  let open Fiber.O in
+  let* p = get_process t in
+  match p with
+  | Error `No_process -> Fiber.return @@ Error (`Msg "No process")
+  | Error (`Msg msg) -> Fiber.return @@ Error (`Msg msg)
+  | Ok process -> (
+    match
+      Scheduler.async process.io_thread (fun () ->
+          Ocamlformat_rpc_lib.format typ process.client)
+    with
+    | Error `Stopped -> Fiber.return @@ Error (`Msg "Thread stopped")
```

Comment:
Is this error message going to be visible to the user? If so,

* Under what circumstances should the user expect this?
* The error message should be more descriptive as to what is happening.

Same for the `"No process"` message.